simples3.go

```diff
@@ -149,12 +152,55 @@ func NewUsingIAM(region string) (*S3, error) {
 		}, securityCredentialsURL, region)
 }
 
+// fetchIMDSToken retrieves an IMDSv2 token from the
+// EC2 instance metadata service. It returns a token
+// only if IMDSv2 is enabled in the EC2 instance metadata
+// configuration, otherwise returns an error.
+func fetchIMDSToken(cl *http.Client) (string, error) {
+	req, err := http.NewRequest(http.MethodPut, imdsTokenURL, nil)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("X-aws-ec2-metadata-token-ttl-seconds", "21600")
+
+	resp, err := cl.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("failed to fetch IMDSv2 token: %s", resp.Status)
+	}
+
+	token, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return string(token), nil
+}
+
 // fetchIAMData fetches the IAM data from the given URL.
 // In case of a normal AWS setup, baseURL would be securityCredentialsURL.
 // You can use this method, to manually fetch IAM data from a custom
 // endpoint and pass it to SetIAMData.
 func fetchIAMData(cl *http.Client, baseURL string) (IAMResponse, error) {
-	resp, err := cl.Get(baseURL)
+	var token string
+	token, err := fetchIMDSToken(cl)
```

Comment:
This signature and error check doesn't make sense while reading as the error is just equivalent to useIMDSv2. 

May I propose we change this to `fetchIMDSToken(client) (token, useIMDSv2, error)`

This way it is cleaner and makes the errors actually usable if any.

Or, we could also use named errors and check with errors.Is() ?

Reply:
https://stackoverflow.com/questions/64595032/how-to-tell-what-version-of-instance-metadata-serviceimds-ec2-instance-is-usin

Another alternative could be to follow this process. Check for 401 and then try with IMDSv2.

Reply:
Yeah, the error is kind of pointless in this case, since we don't print it either. We could get rid of the error and instead just return a boolean value specifying whether IMDSv2 can be used. Would that be okay?

Reply:
I think we should still return an error, just in case this is used later on.

fetchIMDSToken(client) (token, useIMDSv2, error) <- should be okay

useIMDSv2 can be false if we get 401. Any other kind of error, we can send error.